pkg/skaffold/jib/jib.go

```diff
@@ -17,32 +17,114 @@ limitations under the License.
 package jib
 
 import (
+	"encoding/json"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"sort"
+	"time"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 	"github.com/karrick/godirwalk"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-func getDependencies(cmd *exec.Cmd) ([]string, error) {
-	stdout, err := util.RunCmdOut(cmd)
-	if err != nil {
-		return nil, err
+// filesTemplate contains fields parsed from Jib's JSON output
+type filesTemplate struct {
+	Build  []string
+	Inputs []string
+	Ignore []string
+}
+
+// filesLists contains cached build/input dependencies
+type filesLists struct {
+	WatchedBuildFiles []string
+	WatchedInputFiles []string
+	BuildFileTimes    map[string]time.Time
+}
+
+// watchedFiles maps from project name to watched files
+var watchedFiles = map[string]filesLists{}
+
+func getDependencies(cmd *exec.Cmd, projectName string) ([]string, error) {
+	if len(watchedFiles[projectName].WatchedInputFiles) == 0 && len(watchedFiles[projectName].WatchedBuildFiles) == 0 {
+		// Make sure build file modification time map is setup
+		if watchedFiles[projectName].BuildFileTimes == nil {
+			watched := watchedFiles[projectName]
+			watched.BuildFileTimes = make(map[string]time.Time)
+			watchedFiles[projectName] = watched
+		}
+
+		// Refresh dependency list if empty
+		if err := refreshDependencyList(cmd, projectName); err != nil {
+			return nil, err
+		}
+	} else {
+		// Refresh dependency list if any build definitions have changed
+		for _, buildFile := range watchedFiles[projectName].WatchedBuildFiles {
+			info, err := os.Stat(buildFile)
+			if err != nil {
+				return nil, err
+			}
+			if val, ok := watchedFiles[projectName].BuildFileTimes[buildFile]; !ok || info.ModTime() != val {
+				if err := refreshDependencyList(cmd, projectName); err != nil {
+					return nil, err
+				}
+			}
+		}
 	}
 
-	cmdDirInfo, err := os.Stat(cmd.Dir)
+	files := append(watchedFiles[projectName].WatchedBuildFiles, watchedFiles[projectName].WatchedInputFiles...)
+	sort.Strings(files)
+	return files, nil
+}
+
+func refreshDependencyList(cmd *exec.Cmd, projectName string) error {
+	stdout, err := util.RunCmdOut(cmd)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	// Parses stdout for the dependencies, one per line
 	lines := util.NonEmptyLines(stdout)
+	for i := range lines {
+		if lines[i] == "BEGIN JIB JSON" {
+			// Found Jib JSON header, next line is the JSON
+			var filesOutput filesTemplate
+			if err := json.Unmarshal([]byte(lines[i+1]), &filesOutput); err != nil {
+				return err
+			}
+
+			// Walk the files in each list and filter out ignores
+			files := watchedFiles[projectName]
+			files.WatchedInputFiles, err = files.walkFiles(&filesOutput, false)
+			if err != nil {
+				return err
+			}
+			files.WatchedBuildFiles, err = files.walkFiles(&filesOutput, true)
```

Comment:
Since `walkFiles` is a method of `files`, can't it do both the build and input files at once?  It seems odd to have this method assign the result of an object-method to an object-field.